packages/storage/src/implementation/list.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Documentation for the listOptions and listResult format
+ */
+import { AuthWrapper } from './authwrapper';
+import { Location } from './location';
+import * as json from './json';
+import * as type from './type';
+import { ListResult } from '../list';
+
+/**
+ * Represents the simplified object metadata returned by List API.
+ * Other fields are filtered because list does not grant permission to get metadata.
+ */
+interface ListMetadataResponse {
+  name: string;
+  bucket: string;
+}
+
+/**
+ * Represents the JSON response of List API.
+ */
+interface ListResultResponse {
+  prefixes: string[];
+  items: ListMetadataResponse[];
+  nextPageToken?: string;
+}
+
+const MAX_RESULTS_KEY = 'maxResults';
+const PAGE_TOKEN_KEY = 'pageToken';
+const PREFIXES_KEY = 'prefixes';
+const ITEMS_KEY = 'items';
+
+function fromBackendResponse(
+  authWrapper: AuthWrapper,
+  resource: ListResultResponse
+): ListResult {
+  const listResult: ListResult = {
+    prefixes: [],
+    items: [],
+    nextPageToken: resource['nextPageToken']
+  };
+  if (resource[PREFIXES_KEY]) {
+    for (const path of resource[PREFIXES_KEY]) {
+      const pathWithoutTrailingSlash = path.replace(/\/$/, '');
+      const reference = authWrapper.makeStorageReference(
+        new Location(authWrapper.bucket(), pathWithoutTrailingSlash)
+      );
+      listResult.prefixes.push(reference);
+    }
+  }
+
+  if (resource[ITEMS_KEY]) {
+    for (const item of resource[ITEMS_KEY]) {
+      const reference = authWrapper.makeStorageReference(
+        new Location(authWrapper.bucket(), item['name'])
+      );
+      listResult.items.push(reference);
+    }
+  }
+  return listResult;
+}
+
+export function fromResposneString(
+  authWrapper: AuthWrapper,
+  resourceString: string
+): ListResult | null {
+  const obj = json.jsonObjectOrNull(resourceString);
+  if (obj === null) {
+    return null;
+  }
+  const resource = obj as ListResultResponse;
+  return fromBackendResponse(authWrapper, resource);
+}
+
+export function listOptionsValidator(p: any) {
+  const validType = p && type.isObject(p);
+  if (!validType) {
+    throw 'Expected ListOptions object.';
+  }
+  for (const key in p) {
+    if (key === MAX_RESULTS_KEY) {
+      if (!type.isInteger(p[MAX_RESULTS_KEY]) || p[MAX_RESULTS_KEY] <= 0) {
```

Comment:
Where do we validate that this number is less than or equal to 1000? Is that still missing?

Reply:
Sorry, just added that and a test.